test/unit/serializer.cpp

```diff
@@ -250,6 +255,88 @@ struct serializer_test
         f(s);
     };
 
+    template <class F>
+    void
+    check_stream(
+        core::string_view headers,
+        core::string_view body,
+        F f)
+    {
+        response res(headers);
+
+        serializer sr(1024);
+        auto stream = sr.start_stream(res);
+
+        std::size_t const n = 128;
+        BOOST_TEST(body.size() % n == 0);
+
+        std::vector<char> s;
+        while(! sr.is_done() )
+        {
+            if(! body.empty() )
+            {
+                auto mbs = stream.prepare(n);
+                auto bs = buffers::buffer_size(mbs);
+                BOOST_TEST_EQ(bs, n);
+
+                if( bs > body.size() )
+                    bs = body.size();
+
+                buffers::buffer_copy(
+                    mbs,
+                    buffers::const_buffer(
+                        body.data(), bs));
+
+                stream.commit(bs);
+                body.remove_prefix(bs);
+            }
+
+            if( body.empty() )
+            {
+                stream.close();
+                BOOST_TEST_THROWS(
+                    stream.close(), std::logic_error);
+            }
+
+            auto mcbs = sr.prepare();
+            BOOST_TEST(!mcbs.has_error());
+
+            auto cbs = mcbs.value();
+            auto size = buffers::buffer_size(cbs);
+            BOOST_TEST_GT(size, 0);
+
+            // we have the prepared buffer sequence
+            // representing the serializer's output but we
+            // wish to emulate a user consuming it using a
+            // smaller, fixed-size buffer
```

Comment:
This is not good enough, you want to test breaking a body of size N into two pieces of size M and (N-M), where M > 0, for all M. Breaking a body of size N into N/16 rounded down pieces plus one piece of size N mod 16 does not cover enough cases.

Reply:
I'm not sure I understand what you mean here.

One thing that I think we could do is use `std::rand()` to help us generate chunk sizes for the `stream::prepare(n)` call.

Otherwise, what this piece of code is really doing is `consume`'ing the serializer's output in small increments which winds up stress testing the code pretty well. Which is why I'm not sure about breaking a body of size N into two.

The only times we're breaking a body up is in the above `if(! body.empty() )` loop and it'd be a good test to break it up into a random series of sized chunks.

Reply:
random numbers is the antithesis of rigorous testing